### PR TITLE
fix: duplication on adding policies to db

### PR DESCRIPTION
**Pull request by ndodanli**

### CODE REVIEW

File: adapter.go [hunk 1]
```diff
@@ -358,8 +358,18 @@ func (a *Adapter) AddPolicy(sec string, ptype string, rule []string) error {
 func (a *Adapter) AddPolicies(sec string, ptype string, rules [][]string) error {
 	_, err := a.engine.Transaction(func(tx *xorm.Session) (interface{}, error) {
 		for _, rule := range rules {
+			var err error
+			var isLineExist bool
 			line := a.genPolicyLine(ptype, rule)
-			_, err := tx.InsertOne(line)
+			// check if the rule already exists, if it does, skip it
+			isLineExist, err = tx.Get(line)
```
Comment (hsluoyz) [inline adapter.go:365]: Can't you get all rules in one time? Geting one by one is too slow

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: adapter.go (hunk 1 above)
Comment (ndodanli) [reply adapter.go]: Performance increased

**Review by hsluoyz — APPROVED** [review]

**Review by hsluoyz — CHANGES_REQUESTED** [review]
@ndodanli fix the CI:

image

**Comment by ndodanli** [comment]
There is a column naming difference between databases that i think xorm column naming is responsible for this. For example, 'pgtype' for mysql and 'pg_type' for postgres. I determine by checking driver name but setting a fixed column name for all databases is better, but setting fixed column names is not up to me.

Also i saw a usage for directly calling "pgtype" on `UpdateFilteredPolicies()(adapter.go:550)->queryString()(adapter.go:643)` and this can cause problems on postgres. You might want to check on this.

**Comment by hsluoyz** [comment]
@ndodanli this PR is becoming more complicated, even involving home-made column names now. And the tests still failed. I have no confidence for this PR. So closed here. You can do de-duplication on your side anyway.

**Pull request closed without merging**